drv/spi-msg/src/lib.rs

```diff
@@ -0,0 +1,166 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Client API for SP to RoT messages over SPI.
+
+#![no_std]
+
+use derive_idol_err::IdolError;
+use zerocopy::{AsBytes, ByteSliceMut, FromBytes, LayoutVerified, Unaligned};
+//use userlib::*;
+use userlib::{sys_send, FromPrimitive};
+
+#[derive(Copy, Clone, Debug, FromPrimitive, PartialEq, IdolError)]
+pub enum MsgError {
+    /// Transfer size is outside of maximum and minimum lenghts for message type.
+    BadTransferSize = 1,
+
+    /// Server restarted
+    ServerRestarted = 2,
+
+    /// FIFO overflow/underflow
+    FlowError = 3,
+
+    /// Unsupported protocol version
+    UnsupportedProtocol = 4,
+
+    /// Unknown message
+    BadMessageType = 5,
+
+    /// Transfer size is outside of maximum and minimum lenghts for message type.
+    BadMessageLength = 6,
+
+    /// Error from Spi
+    SpiServerError = 7,
+}
+
+/// Protocol version
+pub const SPI_MSG_IGNORE: u8 = 0; // To be ignored
+pub const SPI_MSG_VERSION: u8 = 1; // Supported message format
+
+/// SPI Message types will allow for multiplexing and forward compatibility.
+#[derive(
+    Copy, Clone, PartialEq, Eq, Debug, userlib::FromPrimitive, AsBytes,
+)]
+#[repr(u8)]
+pub enum MsgType {
+    Invalid = 0,
+    ErrorRsp = 1,
+    EchoReq = 2,
+    EchoRsp = 3,
+    StatusReq = 4,
+    StatusRsp = 5,
+    SprocketsReq = 6,
+    SprocketsRsp = 7,
+    Unknown = 0xff,
+}
+
+impl From<u8> for MsgType {
```

Comment:
You're already deriving `FromPrimitive` – do you really need `From` as well?

If you do, you can implement it in terms of `FromPrimitive`, e.g. `MsgType::from_u8(msgtype).unwrap_or(MsgType::Unknown)`

Reply:
I was able to get rid of the issues I had by using accessors outside of sprot-api (nee spi-msg) and using u8 as the internal type.